File: A&DS/mergesort.py
def merge_sort(arr):
    print("Splitting:", arr)
    if len(arr) <= 1:
        return arr
    
    # Divide the array into two halves
    mid = len(arr) // 2
    left_half = arr[:mid]
    right_half = arr[mid:]
    
    # Recursively sort each half
    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)
    
    # Merge the sorted halves
    return merge(left_half, right_half)

def merge(left, right):
    print("Merging:", left, right)
    merged = []
    left_index, right_index = 0, 0
    
    # Compare elements from both lists and add the smaller one to the merged list
    while left_index < len(left) and right_index < len(right):
        if left[left_index] < right[right_index]:
            merged.append(left[left_index])
            left_index += 1
        else:
            merged.append(right[right_index])
            right_index += 1
    
    # Add any remaining elements from the left and right lists
    merged.extend(left[left_index:])
    merged.extend(right[right_index:])
    
    print("Merged:", merged)
    return merged

# Example usage:
arr = [3, 1, 4, 1, 5, 9, 2, 6]
sorted_arr = merge_sort(arr)
print("Sorted array:", sorted_arr)

File: Year2/A&DS/mergesort.c
#include <stdio.h>
void merge(int arr[], int left, int mid, int right) {
  int i, j, k;

  int left_size = mid - left + 1;
  int right_size = right - mid;
  int left_half[left_size];
  int right_half[right_size];
  for (i = 0; i < left_size; i++) {
    left_half[i] = arr[left + i];
  }
  for (j = 0; j < right_size; j++) {
    right_half[j] = arr[mid + 1 + j];
  }
  i = 0;
  j = 0;
  k = left;
  while (i < left_size && j < right_size) {
    if (left_half[i] <= right_half[j]) {
      arr[k] = left_half[i];
      i++;
    } else {
      arr[k] = right_half[j];
      j++;
    }
    k++;
  }
  while (j < right_size) {
    arr[k] = right_half[j];
    j++;
    k++;
  }
  while (i < left_size) {
    arr[k] = left_half[i];
    i++;
    k++;
  }
}
void merge_sort(int arr[], int left, int right) {
  if (left < right) {
    int mid = left + (right - left) / 2;
    merge_sort(arr, left, mid);
    merge_sort(arr, mid + 1, right);
    merge(arr, left, mid, right);
  }
}

void print_array(int arr[], int size) {
  int i;
  for (i = 0; i < size; i++) {
    printf("[%d]", arr[i]);
  }
  printf("\n");
}
int main() {
  int arr[] = {3, 1, 4, 1, 5, 9, 2, 6};
  int n = sizeof(arr) / sizeof(arr[0]);
  printf("Given array is \n");
  print_array(arr, n);
  merge_sort(arr, 0, n - 1);
  printf("\nSorted array is \n");
  print_array(arr, n);
  printf("\n");
  return 0;
}
